shared/pow.go

```diff
@@ -0,0 +1,75 @@
+package shared
+
+import (
+	"context"
+	"encoding/binary"
+	"hash"
+
+	"github.com/c0mm4nd/go-ripemd"
+)
+
+// FindSubmitPowNonce finds the nonce that solves the PoW challenge.
+func FindSubmitPowNonce(
+	ctx context.Context,
+	powChallenge, poetChallenge, nodeID []byte,
+	difficulty uint,
+) (uint64, error) {
+	var hash []byte
+	p := NewPowHasher(powChallenge, nodeID, poetChallenge)
+
+	for nonce := uint64(0); ; nonce++ {
+		select {
+		case <-ctx.Done():
+			return 0, ctx.Err()
+		default:
+		}
+
+		hash = p.Hash(nonce, hash[:0])
+
+		if CheckLeadingZeroBits(hash, difficulty) {
+			return nonce, nil
+		}
+	}
+}
+
+type powHasher struct {
+	h     hash.Hash
+	input []byte
+}
+
+func NewPowHasher(inputs ...[]byte) *powHasher {
+	h := &powHasher{h: ripemd.New256(), input: []byte{}}
```

Comment:
NIT: `h.input` does not need to be initialized to append bytes to it 

```suggestion
	h := &powHasher{h: ripemd.New256()}
```